### PR TITLE
[Fixes #10859] Deadlock in HierarchicalTagManager when creating tags

### LINKED ISSUE #10859: Deadlock in HierarchicalTagManager when creating tags
## Actual Behavior

Sometimes the log reports a DB deadlock when creating tags.

## Specifications

  - GeoNode version: 3.3.x, 4.x

## Details

<details>
  <summary>Log `tag_objs.update(existing)` --> `self._fetch_all()` </summary>

```
Traceback (most recent call last):
  File "/usr/src/geonode/geonode/upload/upload.py", line 936, in final_step
    saved_layer = _update_layer_with_xml_info(saved_layer, xml_file, regions, keywords, vals)
  File "/usr/src/geonode/geonode/upload/upload.py", line 1075, in _update_layer_with_xml_info
    saved_layer = utils.KeywordHandler(saved_layer, keywords).set_keywords()
  File "/usr/src/geonode/geonode/upload/utils.py", line 669, in set_keywords
    self._set_free_keyword(keywords)
  File "/usr/src/geonode/geonode/upload/utils.py", line 711, in _set_free_keyword
    self.instance.keywords.add(*keywords)
  File "/usr/src/geonode/geonode/base/models.py", line 478, in add
    tag_objs.update(existing)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 274, in __iter__
    self._fetch_all()
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 1242, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/local/lib/python3.8/site-packages/django/db/models/query.py", line 55, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/usr/local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1142, in execute_sql
    cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.OperationalError: deadlock detected
DETAIL:  Process 19352 waits for ShareLock on transaction 71406198; blocked by process 19188.
Process 19188 waits for ShareLock on transaction 71406498; blocked by process 15015.
Process 15015 waits for AccessExclusiveLock on tuple (23,57) of relation 40986 of database 37725; blocked by process 19352.
HINT:  See server log for query details.
CONTEXT:  while locking tuple (23,57) in relation "base_hierarchicalkeyword"
```
</details>

<details>
  <summary>Log `new_tag_obj = HierarchicalKeyword.add_root(name=new_tag)` </summary>

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/geonode/geonode/upload/upload.py", line 936, in final_step
    saved_layer = _update_layer_with_xml_info(saved_layer, xml_file, regions, keywords, vals)
  File "/usr/src/geonode/geonode/upload/upload.py", line 1075, in _update_layer_with_xml_info
    saved_layer = utils.KeywordHandler(saved_layer, keywords).set_keywords()
  File "/usr/src/geonode/geonode/upload/utils.py", line 669, in set_keywords
    self._set_free_keyword(keywords)
  File "/usr/src/geonode/geonode/upload/utils.py", line 711, in _set_free_keyword
    self.instance.keywords.add(*keywords)
  File "/usr/src/geonode/geonode/base/models.py", line 483, in add
    new_tag_obj = HierarchicalKeyword.add_root(name=new_tag)
  File "/usr/local/lib/python3.8/site-packages/treebeard/mp_tree.py", line 617, in add_root
    return MP_AddRootHandler(cls, **kwargs).process()
  File "/usr/local/lib/python3.8/site-packages/treebeard/mp_tree.py", line 313, in process
    return last_root.add_sibling('sorted-sibling', **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/treebeard/mp_tree.py", line 1098, in add_sibling
    return MP_AddSiblingHandler(self, pos, **kwargs).process()
  File "/usr/local/lib/python3.8/site-packages/treebeard/mp_tree.py", line 436, in process
    self.run_sql_stmts()
  File "/usr/local/lib/python3.8/site-packages/treebeard/mp_tree.py", line 145, in run_sql_stmts
    cursor.execute(sql, vals)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/usr/local/lib/python3.8/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.OperationalError: deadlock detected
DETAIL:  Process 19188 waits for ShareLock on transaction 71406498; blocked by process 15015.
Process 15015 waits for AccessExclusiveLock on tuple (23,57) of relation 40986 of database 37725; blocked by process 19377.
Process 19377 waits for ShareLock on transaction 71406198; blocked by process 19188.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (6,55) in relation "base_hierarchicalkeyword"

```
</details>

### PR DESCRIPTION
**Pull request by mattiagiupponi**
ref #10859 
## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/GeoNode/geonode/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in the documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for contrib modules):
- [x] There is a ticket in https://github.com/GeoNode/geonode/issues describing the issue/improvement/feature (a notable exemption is, changes not visible to end-users)
- [x] The issue connected to the PR must have Labels and Milestone assigned
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[Fixes #<issue_number>] Title of the Issue"
- [ ] New unit tests have been added covering the changes, unless there is an explanation on why the tests are not necessary/implemented
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the QA checks: black geonode && flake8 geonode
- [ ] Commits changing the **settings**, **UI**, **existing user workflows**, or adding **new functionality**, need to include documentation updates
- [ ] Commits adding **new texts** do use gettext and have updated .po / .mo files (without location infos)

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

### CODE REVIEW

**Comment by giohappy** [comment]
@mattiagiupponi the select for update was introduced in https://github.com/GeoNode/geonode/commit/1c22acab2b2580f08debe617a8d5d7f5d1e6964a to fix [#8913](https://github.com/GeoNode/geonode/issues/8913).
Just removing it will reintroduce the problem for which it was introduced....

**Comment by etj** [comment]
@mattiagiupponi I'd add a `.all()` to perform the query as soon as possible

Pls also remember to squash the PR when merging.

File: geonode/base/models.py [hunk 1]
```diff
@@ -397,49 +398,57 @@ def add(self, *tags, through_defaults=None, tag_kwargs=None):
...
-        existing = self.through.tag_model().objects.select_for_update().filter(name__in=str_tags, **tag_kwargs)
-        with transaction.atomic():
-            tag_objs.update(existing)
-            new_ids = set()
-            _new_keyword = str_tags - set(t.name for t in existing)
-            for new_tag in list(_new_keyword):
-                new_tag = escape(new_tag)
+        try:
+            with transaction.atomic():
+                existing = self.through.tag_model().objects.all().filter(name__in=str_tags, **tag_kwargs)
```
Comment (etj) [inline geonode/base/models.py:403]:
```python
self.through.tag_model().objects.filter(name__in=str_tags, **tag_kwargs).all()
```

**Comment by etj** [comment]
@mattiagiupponi the concurrency problem with treebeard is long known and not resolved yet.
I guess that we may want to solve it the hard way:
- create a method that receives a list of keywords to be created
- in the method we create a brand new db connction, and try to inserta all the keywords we can
- if not all the kw have been inserted because of an exception, we call the method again, tha will create a new connection and try to add the kw that are not in the db yet.

We need a decoupled connection bc when a db error happens, the whole current transaction is rolled back,and we risk to lose the layer info that have just been inserted.

**Comment by giohappy** [comment]
I would try to solve the logical problem rather then biting with retries.

We should try isolating the transaction where we  try to add the keyword, so that if it fails we should be able to recover from it.
Something like the following:

```
existing = self.through.tag_model().objects.filter(
            name__in=str_tags, **tag_kwargs
        )
        new_ids = set()
        _new_keywords = str_tags - set(t.name for t in existing)
        for new_tag in list(_new_keywords):
            new_tag = escape(new_tag)
            new_tag_obj = None
            with transaction.atomic():
                try:
                    new_tag_obj = HierarchicalKeyword.add_root(name=new_tag)
                except Exception as e:
                    logger.exception(e)
            # HierarchicalKeyword.add_root didn't return, probably the keyword already exists
            if not new_tag_obj:
                new_tag_obj = HierarchicalKeyword.objects.get(name=new_tag)
            if new_tag_obj:
                tag_objs.add(new_tag_obj)
                new_ids.add(new_tag_obj.id)
            else:
                # Something has gone seriously wrong and we cannot assign the tag to the resource
                logger.exception(e)
```

**Review by giohappy — APPROVED** [review]

**Pull request merged**